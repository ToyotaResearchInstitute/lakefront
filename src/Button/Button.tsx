import { FC } from 'react';
import ButtonVariants from './buttonVariants';
import IconButton from './IconButton';
import {
    ButtonComponentProps,
    COLORS,
    ICON_BUTTON_COLOR,
    ICON_NAMES,
    InvalidButtonColorError,
    shouldUseMappedIcon
} from './buttonUtil';

/**
 * Button Component
 *
 * The Button component takes a color and an alt prop, as well as all native button props.
 * The color prop, which defaults to primary, determines its overall color scheme.
 * The alt prop inverts the color scheme to look better on dark backgrounds.
 * className will properly apply itself, but the CSS may need to be more specific,
 * such as button.MyClass instead of just .MyClass.
 *
 * The button can also include an icon configuration to include an icon. The icon configuration includes
 * a name, an optional tooltip, and an optional position. The name, which defaults to 'help', determines which icon to display along side any
 * provided children. If provided, the tooltip will cause the provided text to be displayed
 * when hovering over the button. The position, which defaults to 'left', determines if
 * the icon precedes or proceeds the provided children.
 */
const Button: FC<ButtonComponentProps> = ({
    alternate = false,
    color = COLORS.PRIMARY,
    children,
    icon = false,
    iconPosition = 'left',
    ...props
}) => {
    // Like in the Icon version, we return a styled component based on the color type
    const ButtonComponent = ButtonVariants[color];

    return (
        <ButtonComponent
            alternate={alternate}
            icon={icon}
            iconPosition={iconPosition}
            {...props}
        >
            {children}
        </ButtonComponent>
    );
};

export default Button;
