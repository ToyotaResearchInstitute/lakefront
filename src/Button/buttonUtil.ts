import { ComponentPropsWithoutRef, ElementType } from 'react';
import { SerializedStyles } from '@emotion/react';

export interface ComponentStyles {
    [key: string]: SerializedStyles
}

export enum COLORS {
    PRIMARY = 'primary',
    SECONDARY = 'secondary',
    DESTRUCTIVE = 'destructive'
}

export const ICON_BUTTON_COLOR: { [key: string]: 'secondary' | 'destructive' } = {
    primary: 'secondary',
    secondary: 'secondary',
    destructive: 'destructive'
};

export const ICON_NAMES = {
    primary: 'add',
    secondary: 'add',
    destructive: 'delete'
};

export class InvalidButtonColorError extends Error {
    color: string | undefined;

    constructor(color: string | undefined) {
        super(`Invalid color '${color}' was provided. Valid colors are: ${Object.values(COLORS).join(', ')}.`);
    }
}

export interface ButtonProps {
    color?: COLORS;
    alternate?: boolean;
    as?: ElementType | keyof JSX.IntrinsicElements;
    icon?: boolean;
    position?: 'left' | 'right';
}

export type ButtonComponentProps = ButtonProps & ComponentPropsWithoutRef<'button'>;

export const shouldUseMappedIcon = (props: ButtonComponentProps): boolean | undefined => {
    return props.icon && !('iconName' in props);
};
