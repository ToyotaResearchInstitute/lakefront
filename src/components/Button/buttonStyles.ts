import styled from '@emotion/styled';

export const StyledDiv = styled.div({
    display: 'inline-block',
    textAlign: 'center'
});

export const IconLabelStyles = styled.div({
    marginTop: 10,
});
