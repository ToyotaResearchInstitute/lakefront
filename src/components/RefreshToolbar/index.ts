import RefreshToolbar, { RefreshToolbarProps } from './RefreshToolbar';

export default RefreshToolbar;
export { RefreshToolbarProps };
