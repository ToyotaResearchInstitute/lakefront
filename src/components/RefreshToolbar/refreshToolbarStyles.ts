import styled from '@emotion/styled';
import Loading from 'src/components/Loading/Loading';

export const ToolbarContainer = styled.div<any>(({ standalone, theme }) => ({
    alignItems: 'center',
    backgroundColor: theme?.colors?.white,
    boxSizing: 'border-box',
    display: 'flex',
    height: '50px',
    zIndex: 1,
    padding: '0 16px',
    ...(standalone) && {
        borderBottom: '1px solid',
        borderColor: theme?.colors?.mercury
    }
}));


export const LogoLoadingSpinner = styled(Loading)(({ width, height, theme }) => ({
    'svg path': {
        fill: 'currentColor',
        width: width,
        height: height
    }
}));

export const RefreshProgressContainer = styled.div<any>(({ theme }) => ({
    alignItems: 'center',
    color: theme?.colors?.pavement,
    display: 'inline-flex',
    fontSize: '13px',
    paddingTop: '3px',
    '@keyframes spin': {
        '0%': { transform: 'rotate(0deg)' },
        '100%': { transform: 'rotate(360deg)' }
    },
    'img': {
        animation: 'spin 2s linear infinite',
        height: '24px',
        width: '24px',
        marginLeft: '5px',
        marginRight: '10px'
    },
    'span': {
        paddingTop: '4px'
    }
}));

export const ToolbarAdditionalContent = styled.div<any>(({ theme }) => ({
    display: 'inline-grid',
    gridTemplateColumns: '1fr minmax(225px, max-content)',
    width: '100%',
    zIndex: 'inherit'
}));

export const ButtonContainer = styled.div<any>(({ theme }) => ({
    display: 'flex',
    alignItems: 'center',
    button: {
        border: 0,
        borderRadius: '50%',
        height: '40px',
        width: '40px',
        minWidth: '40px',
        padding: 0,
        margin: '0 8px',
        'svg': {
            color: theme?.colors?.dolphin,
        },
        '&:hover': {
            backgroundColor: theme?.colors?.selago,

            'svg path': {
                color: theme?.colors?.storm
            }
        },
        '&:disabled': {
            'svg': {
                fill: theme?.colors?.mercury
            }
        }
    }
}));

export const RightSideContainer = styled.div<any>(({ theme }) => ({
    alignItems: 'center',
    color: theme?.colors?.pavement,
    display: 'inline-flex',
    fontSize: '13px',
    paddingTop: '3px',
    justifyContent: 'flex-end'
}));
