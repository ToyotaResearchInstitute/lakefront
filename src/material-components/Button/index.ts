import ButtonComponent, { ButtonProps } from './Buttton';

export { ButtonProps as MaterailButtonProps };
export default ButtonComponent;
