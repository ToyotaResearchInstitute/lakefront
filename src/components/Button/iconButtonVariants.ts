import { css, Theme } from '@emotion/react';
import { lightenDarkenColor } from 'src/styles/stylesUtil';
import { ComponentStyles } from './buttonUtil';

const MAIN_STYLES = {
    border: 'none',
    minWidth: 0,
    height: 48,
    width: 48,
    borderRadius: '100%'
};

interface GetIconStylesProps {
    theme?: Theme;
    alternate?: boolean;
    disabled?: boolean;
}


export const getIconStyles = ({ alternate, disabled, theme }: GetIconStylesProps): ComponentStyles => {
    const primaryBackground = lightenDarkenColor(alternate ? theme?.colors?.white : theme?.colors?.storm, -10);
    const secondaryBackground = disabled ? 'transparent' : lightenDarkenColor(theme?.colors?.white, -10);
    const destructiveBackground = alternate ?
        lightenDarkenColor(theme?.colors?.saturatedRed, -10) :
        theme?.colors?.saturatedRed;

    return {
        primary: css({
            ...MAIN_STYLES,
            ...(!disabled && {
                ':hover': {
                    backgroundColor: primaryBackground
                }
            })
        }),
        secondary: css({
            ...MAIN_STYLES,
            ...(!disabled && {
                ':hover': {
                    backgroundColor: secondaryBackground
                }
            })
        }),
        destructive: css({
            ...MAIN_STYLES,
            ...(!disabled && {
                ':hover': {
                    backgroundColor: destructiveBackground,
                    span: {
                        color: theme?.colors?.white
                    }
                }
            }),
            border: `1px solid ${theme?.colors?.saturatedRed}`
        })
    };
};
