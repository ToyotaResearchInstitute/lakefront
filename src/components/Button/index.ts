import Button from './Button';
import { ButtonComponentProps as ButtonProps } from './buttonUtil'

export { ButtonProps };
export default Button;
