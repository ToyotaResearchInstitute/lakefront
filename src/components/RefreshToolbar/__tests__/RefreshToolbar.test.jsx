import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Button from 'src/components/Button/Button';
import RefreshToolbar from '../RefreshToolbar';

describe('<RefreshToolbar />', () => {
    let count = 0;
    let refreshing = false;
    const handleRefresh = () => {
        refreshing = true;
        count += 1;
    };

    it('renders button as correctly-styled icon', () => {
        const { container } = render((
            <RefreshToolbar
                handleRefresh={handleRefresh}
                isRefreshing={refreshing}
                refreshProgressLabel="Loading..."
                refreshTooltipText="Refresh LogSync File Details"
            />
        ));

        expect(container).toMatchSnapshot();
    });

    it('renders tooltip on hover', () => {
        const { container } = render((
            <RefreshToolbar
                handleRefresh={handleRefresh}
                isRefreshing={refreshing}
                refreshProgressLabel="Loading..."
                refreshTooltipText="Refresh LogSync File Details"
            />
        ));

        userEvent.hover(screen.getByRole('button'));
        expect(container).toMatchSnapshot();
    });

    it('calls the refresh callback', () => {
        const { getByRole } = render((
            <RefreshToolbar
                handleRefresh={handleRefresh}
                isRefreshing={refreshing}
                refreshProgressLabel="Loading..."
                refreshTooltipText="Refresh LogSync File Details"
            />
        ));

        fireEvent.click(getByRole('button'));

        expect(count).toBe(1);
        expect(refreshing).toBeTruthy();
    });

    it('renders custom refresh button', () => {
        const refreshButton = <Button color='primary' icon={true}></Button>;
        const { container } = render((
            <RefreshToolbar
                handleRefresh={handleRefresh}
                isRefreshing={refreshing}
                refreshProgressLabel="Loading..."
                refreshTooltipText="Refresh LogSync File Details"
                refreshButton={refreshButton}
            />
        ));

        expect(container.getElementsByClassName('iconChild')).toBeDefined();
    });

    it('renders right side text', () => {
        const { container } = render((
            <RefreshToolbar
                handleRefresh={() => null}
                refreshTooltipText="Refresh LogSync File Details"
                lastUpdated="11:28:22 AM"
                rightSideText="Last Updated: 11:28:22 AM EST"
            />
        ));
        expect(container).toHaveTextContent('Last Updated: 11:28:22 AM EST');
    });

});
