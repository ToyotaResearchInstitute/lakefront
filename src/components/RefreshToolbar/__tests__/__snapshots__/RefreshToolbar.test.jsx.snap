// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RefreshToolbar /> renders button as correctly-styled icon 1`] = `
<div>
  <div
    class="css-ausj0h"
  >
    <div>
      <button
        class="css-th9i8z"
      >
        <span
          class="css-1uvgx5t"
        >
          <span
            class="iconChild"
          >
            <svg />
          </span>
        </span>
      </button>
    </div>
    <div
      class="css-e55m9z"
    >
      <div
        class="css-1yye06s"
      />
      <div
        class="css-ybfxeh"
      >
        
      </div>
    </div>
  </div>
</div>
`;

exports[`<RefreshToolbar /> renders tooltip on hover 1`] = `
<div>
  <div
    class="css-ausj0h"
  >
    <div>
      <button
        class="css-th9i8z"
      >
        <span
          class="css-1uvgx5t"
        >
          <span
            class="iconChild"
          >
            <svg />
          </span>
        </span>
      </button>
    </div>
    <div
      class="css-e55m9z"
    >
      <div
        class="css-1yye06s"
      />
      <div
        class="css-ybfxeh"
      >
        
      </div>
    </div>
  </div>
</div>
`;
